SeQuant/core/space.hpp

```diff
@@ -417,109 +314,446 @@ class IndexSpace {
+      return space1;
+    }
+    else {
+      bool same_qns = ((space1.qns() == space2.qns()) && (space1.qns() == space3.qns()));
+      if(!same_qns){throw std::invalid_argument("asking for the intersection of spaces with incompatible quantum number attributes.");}
+      auto intersection_attr = space1.type().intersection(space2.type()).intersection(space3.type());
+      IndexSpace intersection_space = find_IndexSpace({intersection_attr,space1.qns()});
+      if(intersection_space == nulltype && intersection_attr != 0){
+        throw std::invalid_argument("The resulting space is not registered in this context. Add this "
+            "space to the registry with a label to use it.");
+      }
+      else{return intersection_space;}
+    }
+  }
+
+  //return the resulting space corresponding to a bitwise unIon between two spaces.
+  // check to see if the resulting space is actually registered.
+  IndexSpace unIon(const IndexSpace &space1,
+                                 const IndexSpace &space2) {
+    if( space1 == space2){
+      return space1;
+    }
+    else{
+      bool same_qns = space1.qns() == space2.qns();
+      if(!same_qns){throw std::invalid_argument("asking for the intersection of spaces with incompatible quantum number attributes.");}
+      auto unIontype = space1.type().unIon(space2.type());
+      IndexSpace unIonSpace = find_IndexSpace({unIontype,space1.qns()});
+      if(unIonSpace == nulltype){
+        throw std::invalid_argument("The resulting space is not registered in this context. Add this "
+            "space to the registry with a label to use it.");
+      }
+      else{return unIonSpace;}
+    }
+  }
+
+  //return the resulting spaces corresponding to a bitwise excluded or between two spaces.
+  // check to see if the resulting space is actually registered.
+  std::vector<IndexSpace> non_overlapping_spaces(const IndexSpace &space1, const IndexSpace &space2)const {
+    auto attributes = space1.attr().excluded_spaces(space2.attr());
+    std::vector<IndexSpace> result;
+    for(int i = 0; i < attributes.size(); i++){
+      auto excluded_space = find_IndexSpace(attributes[i]);
+      if(excluded_space == nulltype){
+        throw std::invalid_argument("The resulting space is not registered in this context. Add this "
+            "space to the registry with a label to use it.");
+      }
+      result.push_back(excluded_space);
+    }
+    return result;
+  }
+
+// do two spaces have non_overlapping spaces
+bool has_non_overlapping_spaces(const IndexSpace &space1, const IndexSpace &space2)const{
+if(space1.type().exclusionary_or(space2.type()).to_int32() == 0) {return false;}
+else{ return true;}
+}
+ IndexSpace nulltype_()const {
+  return nulltype;
+ }
+
+// pure occupied means all states are occupied in a given reference
+bool is_pure_occupied(IndexSpace IS)const{
+ if(IS == nulltype_()){
+    return  false;
+}
+   if(IS.type().to_int32() <= vacuum_occupied().type().to_int32()){
+return true;
```

Comment:
This implies numeric ordering of bit patterns which (at least to me) is not obvious. I would have expected bit patterns to only indicate whether spaces overlap or not. If the numeric value is taken into account like it is here, this needs to be documented (rather visibly, I would say).

Wouldn't it be possible to leverage the overlap mechanic for this by having the user define `vacuum_occupied` as the union of all possible sub-spaces that are occupied in the vacuum? In that case this function could check that there is an intersection between the given spaces.

Reply:
This does imply an ordering. I am assuming that all registries which deal with fermions should be ordered such that occupied type orbitals come before unoccupied type orbitals. I will document this better. I assume this is a reasonable assumption for any practical choice of partitioning, is there a case where this might not be true?

Reply:
> I assume this is a reasonable assumption for any practical choice of partitioning, is there a case where this might not be true?

It might be, but I think we don't really need it. The index space concept would (from my POV) be simpler if it involved only a single concept (bit patterns to describe overlapping spaces) instead of involving a second, unrelated concept (ordering based on the numeric values of those bit patterns).

Reply:
Additionally, this logic can only ever be applied to a single orbital space group as all other groups would necessarily have higher values where the first group (occupied) would necessarily have lower values and thus by this logic be included in the check.
And if the check was to be extended with a lower bound as to not overlap with lower spaces, this would implicitly end up pre-defining  existing index spaces which would make defining nee ones more involved (as they have to evade falling into one of those implicit assumptions)